### PR TITLE
Add support for Plasma Mobile

### LINKED ISSUE #118650: Notifications and other popups are treated like "normal" windows on Plasma Wayland
**Describe the bug**
When running Plasma Wayland,  notifications and krunner pop-downs are treated like first-class windows and therefore take focus and appear in the task manager.

This used to be an issue back in the day (https://bugs.kde.org/show_bug.cgi?id=365107) so this is in all likelyhood something we do (wrong) on our side.
    
**To Reproduce**
1. follow the steps in https://github.com/NixOS/nixpkgs/pull/117102 to get wayland running
2. log in to the plasma wayland session
3. wait for a notification or trigger krunner
4. observe the focus steal (for notifications) and appearance in the task manager

**Expected behavior**
Notifications and similar should not be treated as a regular window and therefore not show in task manager or receive focus (krunner should still have focus of course).

**Notify maintainers**
Cc: @NixOS/qt-kde 

**Metadata**

 - system: `"x86_64-linux"`
 - host os: `Linux 5.11.11, NixOS, 21.05.git.d3ff65233f9M (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.4pre20210326_dd77f71`
 - channels(root): `""`
 - channels(peter): `""`
 - nixpkgs: `/nix/store/w9zhp284bnx3f1ma6mdbsw008jpnq15v-nixpkgs_unstable`

References:
- https://github.com/NixOS/nixpkgs/pull/117102

### PR DESCRIPTION
**Pull request by samueldr**
**:heavy_check_mark: This is pretty much ready for a final review.**

Note that the first chunk of commits is in:

- Be mindful about reviews that should go in #130151.

This includes work form #58370. Thank you @Slabity.

### What this is

This includes the "basic" set of released applications from the Plasma Mobile project. In addition to the shell configuration, and the NixOS configuration to use Plasma Mobile.

### Known issues

Probably not specific to Plasma Mobile

 - ~~Notifications are "full-screen", probably is the same issue as #118650~~ As expected, fixed with the same solution that fixes #118650.
 - ~~Night colors not working #117036~~ Fixed

Probably specific to Plasma Mobile

 - Long clicking the wi-fi icon goes to the main settings menu (I assume it should go to some sort of network configuration utility) (unclear)
 - ~~Forced numeric passwords!!!!!~~ This is an upstream issue.

### Things done

 - Testing with https://github.com/NixOS/mobile-nixos/pull/351, which works with the UEFI VM, *and* on hardware on a Pine64 Pinephone.
 - [A standard NixOS VM](https://gist.github.com/67692b13614a50f0f485e2316bfc7092)
 - A pinephone

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

cc @NixOS/qt-kde

### CODE REVIEW

**Comment by samueldr** [comment]
#125156, which **only** package apps is :heavy_check_mark: ready for reviewing.

**Comment by samueldr** [comment]
With this last force-push:

 - Rebased on top of current #125156 (plasma mobile gear 21.05).
 - Removed `plasma-phone-settings` as it was largely cruft and unneeded. Instead we get an (internal) malleable option set for kdeglobals/kwinrc.

**Comment by samueldr** [comment]
(Not pushed) Tried rebasing on #127389, but the upgrade to 5.22.3 does not seem to fix the notifications issue. Night color untested.

Side-note: the issue might be different than the one for plasma5 on wayland, in the end, as when reworking the `kwinrc` changes, I saw that by having the `Placement=Maximizing` value removed the notifications were not maximized (duh?). So it looks like part of the issue is that the notifications are being maximized by that config, and that *something* that would normally make them not maximized doesn't work with the current packaging.

**Comment by samueldr** [comment]
Rebased,  #139459 and #139537 fixes the issues with the notification sizes.

The other two known issues noted are not breaking. The "wifi long press" issue is a supposition that the action is wrong, without actually verifying on another Plarma Mobile distro. The night light issues are untested on real hardware. Given that #117036 is closed, I figure it's more likely to be fixed. I'm still assuming if it's broken that it would be broken on "full" Plasma with Wayland.

Current status is:

 - blocked on #100057, #139459 and #139537
 - "blocked" on #125156 (which is pretty much ready already)

**Comment by Mindavi** [comment]
I'm currently testing this. Some notes:

- For some reason the brightness starts at 5%, which is too low to actually enable the display (in my case). When connected to a charger it coincidentally _barely_ shows something. For now I workaround this by doing `brightnessctl set 30%` in an ssh connection, but it's far from ideal. Dunno if there's a way to set this by default (or it could be a systemd script that runs after graphics are setup)
- I had an empty password and it couldn't handle that, after changing it to a numeric password it appears to accept it though (guess that's an upstream bug though).
- The 'normal' plasma desktop seems to work fine
- The 'mobile' plasma desktop crashed the first time I tried to use it (lots of KDE services crashed), but the second time it worked! (it might've been an issue that I started a normal desktop and then the mobile desktop again, who knows!)
- When the screen is dimmed it's unclear how it's turned back on (I set the brightness to 30% again and that worked...)
- Megapixels works (although the back camera crashes the app, but that's the same on phosh)
- It doesn't seem super stable (sometimes things take exceedingly long or stuff crashes)

This seems to look better 'out of the box' than phosh, which (without the GTK patches) doesn't go fullscreen 

I just discovered that I was suffering from a lack of disk space, which may have influenced some of these tests. I'll probably run this for a while and report anything special I find.

<details><summary>Some pictures (low quality)</summary>

photo_2021-10-04_19-40-18
photo_2021-10-04_19-40-33
photo_2021-10-04_19-40-37
photo_2021-10-04_19-40-41

</details>

In general, I think this is mostly ready for a first cut. The main concern I have is with the brightness, since that's really quite annoying.

**Comment by samueldr** [comment]
This latest push includes changes from #130151, in addition to the *Plasma Mobile shell* additions.

**Comment by samueldr** [comment]
The most recent push is a rebase on top of current `nixos-unstable`.

File: pkgs/desktops/plasma-5/plasma-phone-components/default.nix [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  postPatch = ''
+    substituteInPlace bin/kwinwrapper.in \
+      --replace @KDE_INSTALL_FULL_LIBEXECDIR@ "${plasma-workspace}/libexec"
+
+    substituteInPlace bin/plasma-mobile.desktop.cmake \
+      --replace @CMAKE_INSTALL_FULL_LIBEXECDIR@ "${plasma-workspace}/libexec"
+  '';
+
+  cmakeFlags = [
+    "-DNIXPKGS_MKDIR=${getBin coreutils}/bin/mkdir"
```
[the remark this answers is not in the document]
Comment (samueldr) [reply pkgs/desktops/plasma-5/plasma-phone-components/default.nix:38]: Looking at the history of things, see 84ddb3da35d053e23ab05bc59f687779a1519734 where this was used at some point for `plasma-workspace`.

I assume either things changed, and it's not necessary anymore for `plasma-phone-components`, or it never was. So the answer to your questions: it's a convention used by @ttuegel to embed binary paths "neatly" into the programs.

**Comment by samueldr** [comment]
Last update made the Plasma Mobile independent from Plasma Desktop. Before it required needlessly enabling *all* of Plasma Desktop.

I am gearing up to [test on Mobile devices](https://github.com/NixOS/mobile-nixos/pull/351) more in depth, since everything "generic" seems to just work at this point.

**Comment by Mindavi** [comment]
I tested with this configuration: https://github.com/Mindavi/pinephone-nixos-config/commit/122413a5246ae6673f5f50a57ad0ba6367034067

In general, seems to work ok. I can now properly login, sddm seems to work fine. It builds on the pinephone (everything that was not substituted from the cache, that is).

I did see the keyboard crash, but I don't expect this is related to packaging. In general it seems to work ok, so not sure why it crashed sometimes.

- Using kasts, I verified that audio is working (after changing audio settings).
- Flashlight does not seem to be working (not sure if related to packaging here).
- For some reason, the green notification led does not go out after booting
- Looks a lot more 'mobile' than gnome that I was running on it before
- Bootup time ~2 mins or so
- King's Cross (terminal emulator) looks slightly weird (letters far apart), may be an issue with King's Cross itself
- Brightness settings work
- Battery indicator works
- Wifi indicator works
- Settings app works
- In general, icons in apps are working

Have not looked at the code otherwise.

**Comment by samueldr** [comment]
> For some reason, the green notification led does not go out after booting

I don't know if there's a service that handles LEDs in Plasma Mobile. The green LED is simply the default state from Linux's initialization.

> For some reason, the green notification led does not go out after booting

Could be missing fonts.

With all the hardware working (except for flashlight / notification LEDs), it's looking good I guess.

A thing to do would be checking if upstream manages LEDs, and how, and why it doesn't work if so.

**Comment by Mindavi** [comment]
I cannot find any documentation on notification LEDs in plasma mobile, and the UI also doesn't seem to expose anything about it. So I think nothing is done with it.

I've been using this a little bit since I reviewed this, and haven't found any major issues otherwise. I think this is good to go in.

**Comment by samueldr** [comment]
rebased, fix build issue with update to 5.23.3.

@Mindavi do you want to / can you test again?

I'll build and test again for `pine64-pinephone` too, but it worked with the UEFI VM.

**Comment by Mindavi** [comment]
I'd like to, but I gave my pinephone to a friend to try it out, they are still waiting for their librem 5. Hope they can get some experience before they get it (if ever....). I'll probably get it back somewhere in the following weeks, but didn't really plan a date or anything.

So I'll not be able to test soon, unfortunately.

**Comment by Thra11** [comment]
I can try this out on my pinephone this weekend if you'd like anything specific checking or just another person to test it. It's updating now...

**Review by ttuegel — COMMENTED** [review]
The changes in the NixOS module look good to me! :+1:

**Comment by samueldr** [comment]
@Mindavi @Thra11: I didn't need anything specific to be tested. I tried it on my own unit since this is now totally ready for merging. Another pair of eyes, one last time, would have been nice, but not required.

Thank you @ttuegel for looking!

**Comment by samueldr** [comment]
- [x] Verified the nixos tests for plasma5 still pass as expected

**Comment by Thra11** [comment]
Seems to be basically functional, much better than last time I tested it. Lots of weird and wonderful bugs and quirks, but nothing that's obviously related to plasma mobile packaging rather than upstream bugs. Good to see this merged at last!

**Pull request merged**